### PR TITLE
replace (DE)ALLOCATE_LOCAL by malloc/free

### PR DESCRIPTION
**Pull request by uli42**
This is basically a backport of the following commits + replacing
xalloc/xfree by malloc/free. Fixes ArcticaProject/nx-libs#358.

   commit 2761c103311a1160bc483fd0367d654733df8598
   Author: Daniel Stone <daniel@fooishbar.org>
   Date:   Mon Nov 5 14:03:26 2007 +0000

       OS: Remove usage of alloca

       Replace with heap allocations.

   commit 5e363500c86042c394595e1a6633581eb8fcd1bb
   Author: Daniel Stone <daniel@fooishbar.org>
   Date:   Mon Nov 5 14:38:28 2007 +0000

       OS: Remove ALLOCATE_LOCAL from os.h

       Remove ALLOCATE_LOCAL_FALLBACK and DEALLOCATE_LOCAL_FALLBACK from os.h, and
       remove the include of Xalloca.h as well.

### CODE REVIEW

File: nx-X11/programs/Xserver/mi/mibank.c [hunk 1]
```diff
@@ -175,8 +175,8 @@ typedef struct _miBankQueue
         (*pScreenPriv->BankInfo.SetDestinationBank)(pScreen, (_no)) - \
         (pScreenPriv->BankInfo.BankSize * (_no)))
 
-#define ALLOCATE_LOCAL_ARRAY(atype, ntype) \
-    (atype *)ALLOCATE_LOCAL((ntype) * sizeof(atype))
+#define malloc_ARRAY(atype, ntype) \
+    (atype *)malloc((ntype) * sizeof(atype))
```
Comment (sunweaver) [inline nx-X11/programs/Xserver/mi/mibank.c]: Here search+replace failed, I first thought. How does upstream address this code change? Your approach feels like an artifact.... I looked at X.org xserver code and they actually have xalloc_ARRAY everywhere, but later on drop mibank.c completely. I am aware of an issue report for that. So let's leave malloc_ARRAY as is for now, although it really feels like a search+replace artifact.

File: nx-X11/programs/Xserver/Xext/xf86bigfont.c [hunk 2]
```diff
@@ -638,12 +638,12 @@ ProcXF86BigfontQueryFont(
 	        + (nCharInfos+1)/2 * 2 * sizeof(CARD16)
 	      : 0);
 	xXF86BigfontQueryFontReply* reply =
-	   (xXF86BigfontQueryFontReply *) ALLOCATE_LOCAL(rlength);
+	   (xXF86BigfontQueryFontReply *) calloc(1, rlength);
```
Comment (sunweaver) [inline nx-X11/programs/Xserver/Xext/xf86bigfont.c]: Why do you use calloc() here? Uninitialized values issue? If so, I guess, there are more malloc()s in this PR that should possibly become calloc()s?

Comment (sunweaver) [reply nx-X11/programs/Xserver/Xext/xf86bigfont.c]: In X.org xserver, there is commit 

```
commit 0b7b89fbac0b3865b2cf51295c68a5f4c7523f28
Author: Adam Jackson <ajax@redhat.com>
Date:   Mon Oct 6 14:19:30 2008 -0400

    xalloc+bzero -> xcalloc

```
... but it does not mean that X.org xserver had enough bzero()s in place, when they dropped (DE)ALLOCATE_LOCAL.

File: nx-X11/programs/Xserver/mi/mispans.c [hunk 3]
```diff
@@ -531,7 +531,7 @@ void miFillUniqueSpanGroup(pDraw, pGC, spanGroup)
 	free(points);
 	free(widths);
 	free(yspans);
-	free(ysizes);		/* use (DE)ALLOCATE_LOCAL for these? */
+	free(ysizes);		/* use (DE)malloc for these? */
```
Comment (sunweaver) [inline nx-X11/programs/Xserver/mi/mispans.c]: manual fix in comment required!

File: nx-X11/programs/Xserver/record/set.c [hunk 4]
```diff
@@ -71,8 +71,8 @@ typedef int Bool;
 
 typedef unsigned short CARD16;
 
-#define ALLOCATE_LOCAL malloc
-#define DEALLOCATE_LOCAL free
+#define malloc malloc
+#define free free
```
Comment (sunweaver) [inline nx-X11/programs/Xserver/record/set.c]: Drop these two defines in Xserver/record/set.c, they make no sense anymore.

**Review by sunweaver — CHANGES_REQUESTED** [review]
@uli42: I left some comments, some require a fix-up.

All-in-all, I wonder if we should not use calloc() more often rather than malloc() (and thus, divert from what X.org did at the time of those two original commits).

File: nx-X11/programs/Xserver/mi/mibank.c (hunk 1 above)
Comment (uli42) [reply nx-X11/programs/Xserver/mi/mibank.c]: Yes, it is an artifact but it is like upstream handles it. Once mibank will be gone that will vanish anyway. So maybe cherrypicking Ionic's backport of dropping mibank should be done now.

File: nx-X11/programs/Xserver/record/set.c (hunk 4 above)
Comment (uli42) [reply nx-X11/programs/Xserver/record/set.c]: Oh, good point, have not seen these!

File: nx-X11/programs/Xserver/Xext/xf86bigfont.c (hunk 2 above)
Comment (uli42) [reply nx-X11/programs/Xserver/Xext/xf86bigfont.c]: I had added a missing memset here. When replacing ALLOCATE_LOCAL I noticed I can remove the memset again and just use calloc.

calloc is better, right, but keep in mind that a calloc is slower, also.

**Comment by sunweaver** [comment]
On  Fr 03 Mär 2017 11:41:51 CET, Ulrich Sibiller wrote:

> uli42 commented on this pull request.
>
>
>
>> @@ -175,8 +175,8 @@ typedef struct _miBankQueue
>          (*pScreenPriv->BankInfo.SetDestinationBank)(pScreen, (_no)) - \
>          (pScreenPriv->BankInfo.BankSize * (_no)))
>
> -#define ALLOCATE_LOCAL_ARRAY(atype, ntype) \
> -    (atype *)ALLOCATE_LOCAL((ntype) * sizeof(atype))
> +#define malloc_ARRAY(atype, ntype) \
> +    (atype *)malloc((ntype) * sizeof(atype))
>
> Yes, it is an artifact but it is like upstream handels that. Once  
> mibank will be gone the will vanish anyway. So maybe cherrypicking  
> Ionic's backport of dropping mibank should be done now.

I have just cherry-picked that commit and now mibank.[ch] are gone.

Mike
-- 

DAS-NETZWERKTEAM
mike gabriel, herweg 7, 24357 fleckeby
mobile: +49 (1520) 1976 148
landline: +49 (4354) 8390 139

GnuPG Fingerprint: 9BFB AEE8 6C0A A5FF BF22  0782 9AF4 6B30 2577 1B31
mail: mike.gabriel@das-netzwerkteam.de, http://das-netzwerkteam.de

**Comment by sunweaver** [comment]
On  Fr 03 Mär 2017 11:54:25 CET, Ulrich Sibiller wrote:

> uli42 commented on this pull request.
>
>
>
>> @@ -638,12 +638,12 @@ ProcXF86BigfontQueryFont(
>  	        + (nCharInfos+1)/2 * 2 * sizeof(CARD16)
>  	      : 0);
>  	xXF86BigfontQueryFontReply* reply =
> -	   (xXF86BigfontQueryFontReply *) ALLOCATE_LOCAL(rlength);
> +	   (xXF86BigfontQueryFontReply *) calloc(1, rlength);
>
> I had added a missing memset here. When replacing ALLOCATE_LOCAL I  
> noticed I can remove the memset again and just use calloc.
>
> calloc is better, right, but keep in mind that a calloc is slower, also.

yes. Ok. So let's for valgrind to become unhappy again. And then  
filter issues out manually.

Mike
-- 

DAS-NETZWERKTEAM
mike gabriel, herweg 7, 24357 fleckeby
mobile: +49 (1520) 1976 148
landline: +49 (4354) 8390 139

GnuPG Fingerprint: 9BFB AEE8 6C0A A5FF BF22  0782 9AF4 6B30 2577 1B31
mail: mike.gabriel@das-netzwerkteam.de, http://das-netzwerkteam.de

File: nx-X11/programs/Xserver/Xext/xf86bigfont.c (hunk 2 above)
Comment (uli42) [inline nx-X11/programs/Xserver/Xext/xf86bigfont.c]: done

File: nx-X11/programs/Xserver/mi/mispans.c (hunk 3 above)
Comment (uli42) [reply nx-X11/programs/Xserver/mi/mispans.c]: done

File: nx-X11/programs/Xserver/record/set.c (hunk 4 above)
Comment (uli42) [reply nx-X11/programs/Xserver/record/set.c]: done

**Comment by uli42** [comment]
ok, one more round, please

**Review by sunweaver — APPROVED** [review]

**Pull request merged**